src/ray/gcs/gcs_server/gcs_resource_report_poller.cc

```diff
@@ -0,0 +1,148 @@
+
+namespace ray {
+namespace gcs {
+
+GcsResourceReportPoller::GcsResourceReportPoller(
+    std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+    std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool,
+    std::function<void(const rpc::ResourcesData &)> handle_resource_report,
+    std::function<int64_t(void)> get_current_time_milli,
+    std::function<void(
+        const rpc::Address &, std::shared_ptr<rpc::NodeManagerClientPool> &,
+        std::function<void(const Status &, const rpc::RequestResourceReportReply &)>)>
+        request_report)
+    : ticker_(polling_service_),
+      max_concurrent_pulls_(RayConfig::instance().gcs_max_concurrent_resource_pulls()),
+      inflight_pulls_(0),
+      gcs_resource_manager_(gcs_resource_manager),
+      raylet_client_pool_(raylet_client_pool),
+      handle_resource_report_(handle_resource_report),
+      get_current_time_milli_(get_current_time_milli),
+      request_report_(request_report),
+      poll_period_ms_(RayConfig::instance().gcs_resource_report_poll_period_ms()) {}
+
+GcsResourceReportPoller::~GcsResourceReportPoller() { Stop(); }
+
+void GcsResourceReportPoller::Start() {
+  polling_thread_.reset(new std::thread{[this]() {
+    SetThreadName("resource_report_poller");
+    boost::asio::io_service::work work(polling_service_);
+
+    polling_service_.run();
+    RAY_LOG(DEBUG) << "GCSResourceReportPoller has stopped. This should only happen if "
+                      "the cluster has stopped";
+  }});
+  ticker_.RunFnPeriodically([this] { Tick(); }, 100);
+}
+
+void GcsResourceReportPoller::Stop() {
+  if (polling_thread_ != nullptr) {
+    // TODO (Alex): There's technically a race condition here if we start and stop the
+    // thread in rapid succession.
+    polling_service_.stop();
+    if (polling_thread_->joinable()) {
+      polling_thread_->join();
+    }
+  }
+}
+
+void GcsResourceReportPoller::HandleNodeAdded(
+    const std::shared_ptr<rpc::GcsNodeInfo> &node_info) {
+  absl::MutexLock guard(&mutex_);
+
+  auto state = std::make_shared<PullState>();
+  state->node_id = NodeID::FromBinary(node_info->node_id());
+
+  const NodeID &node_id = state->node_id;
+
+  RAY_CHECK(!nodes_.count(node_id)) << "Node with id: " << node_id << " was added twice!";
+
+  state->address.set_raylet_id(node_info->node_id());
+  state->address.set_ip_address(node_info->node_manager_address());
+  state->address.set_port(node_info->node_manager_port());
+
+  state->last_pull_time = -1;
+  state->next_pull_time = get_current_time_milli_();
+
+  nodes_[node_id] = state;
+  to_pull_queue_.push_front(state);
+  RAY_LOG(DEBUG) << "Node was added with id: " << node_id;
+
+  polling_service_.post([this]() { TryPullResourceReport(); });
+}
+
+void GcsResourceReportPoller::HandleNodeRemoved(
+    const std::shared_ptr<rpc::GcsNodeInfo> &node_info) {
+  NodeID node_id = NodeID::FromBinary(node_info->node_id());
+  {
+    absl::MutexLock guard(&mutex_);
+    RAY_CHECK(false);
```

Comment:
?

Reply:
Oh dear... a remanent of debugging... and also a sign i should add a unit test for this code path